include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -202,101 +202,141 @@ struct __init_processing
-                __res = __binary_op(__res, __unary_op(__adjusted_global_id + __stride * __i, __acc...));
+            _Res __res = vectorized_reduction_first<_Res>(__adjusted_global_id, __acc...);
+            for (_Size __i = 1; __i < __no_vec_ops; ++__i)
+            {
+                const _Size __base_idx = __adjusted_global_id + __i * __stride;
+                if (__base_idx + __vec_size_minus_one < __adjusted_n)
+                    vectorized_reduction_remainder(__base_idx, __res, __acc...);
+                else if (__base_idx < __adjusted_n)
+                {
+                    for (_Size __idx = 0; __idx < __vec_size_minus_one; ++__idx)
+                    {
+                        if (__base_idx + __idx >= __adjusted_n)
+                            break;
+                        __res = __binary_op(__res, __unary_op(__base_idx + __idx, __acc...));
+                    }
+                }
+                else
+                    break;
+            }
             __local_mem[__local_idx] = __res;
+            return;
         }
-        else if (__adjusted_global_id < __adjusted_n)
+        // Scalar remainder
+        if (__adjusted_global_id < __adjusted_n)
         {
-            const _Size __items_to_process =
-                std::max(((__adjusted_n - __adjusted_global_id - 1) / __stride) + 1, static_cast<_Size>(0));
-            typename _AccLocal::value_type __res = __unary_op(__adjusted_global_id, __acc...);
-            for (_Size __i = 1; __i < __items_to_process; ++__i)
-                __res = __binary_op(__res, __unary_op(__adjusted_global_id + __stride * __i, __acc...));
+            _Res __res = __unary_op(__adjusted_global_id, __acc...);
+            for (_Size __i = 1; __i < __vec_size_minus_one; ++__i)
+            {
+                const _Size __base_idx = __adjusted_global_id + __i;
+                if (__base_idx >= __adjusted_n)
+                    break;
+                __res = __binary_op(__res, __unary_op(__base_idx, __acc...));
+            }
             __local_mem[__local_idx] = __res;
+            return;
         }
-    }
-
-    // For non-SPIR-V targets, we check if the operator is commutative before selecting the appropriate codepath.
-    // On SPIR-V targets, the sequential implementation with the non-commutative operator is currently more performant.
-    static constexpr bool __use_nonseq_impl = !oneapi::dpl::__internal::__is_spirv_target_v && _Commutative::value;
-
-    template <typename _NDItemId, typename _Size, typename _AccLocal, typename... _Acc>
-    inline void
-    operator()(const _NDItemId& __item_id, const _Size& __n, const _Size& __global_offset, const _AccLocal& __local_mem,
-               const _Acc&... __acc) const
-    {
-        if constexpr (__use_nonseq_impl)
-            return nonseq_impl(__item_id, __n, __global_offset, __local_mem, __acc...);
-        else
-            return seq_impl(__item_id, __n, __global_offset, __local_mem, __acc...);
+        return;
```

Comment:
Can the `return` be removed? We are at the end of the function.

Reply:
Done.